content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md

```diff
@@ -0,0 +1,152 @@
+
+We can find plenty of articles on the internet explaining how to write null-safe code. Null-safety ensures that we have added proper checks in the code to **guarantee the object reference cannot be null or possible safety measures are taken beforehand when any such situation arises**.
+
+Since `NullPointerException` is a runtime exception, it would be hard to figure out such cases during code compilation. Java's type system does not have a way to quickly eliminate the dangerous null object references. 
+
+But, what if I tell you there are helpful approaches to write null-safe code. In particular, if you are using the Spring Framework. Wouldn't that be great news?
+
+Luckily, Spring Framework offers some annotations to solve this exact problem. 
+In this article, we will learn how to use these annotations to write null-safe code using [Spring Boot](https://reflectoring.io/categories/spring-boot/).
+
+## Null-Safety Annotations in Spring
+
+Under the `org.springframework.lang` Spring core package, there are 4 such annotations, `@NonNull`, `@NonNullFields`, `@Nullable`, and `@NonNullApi`. These are [meta-annotations](https://dzone.com/articles/what-are-meta-annotations-in-java) that use the famous (even though dormant at present) [JSR-305](https://jcp.org/en/jsr/detail?id=305) annotations underneath. 
+
+**Popular IDEs like [Eclipse](https://www.eclipse.org/ide/) and [IntelliJ IDEA](https://www.jetbrains.com/idea/) can understand these annotations.** It can warn the developers of potential issues during compile time. 
+
+We are going to use IntelliJ IDEA in this tutorial. Let us find out more with some code examples.
+
+To create the base project, we have taken help from the [Spring Initializr](https://start.spring.io/). The Spring Boot starter is all we need, no need to add any extra dependencies.
+
+## `@NonNull`
+
+Let's use a plain `Employee` class to explain the annotations:
+
+```java
+package io.reflectoring.nullsafety;
+
+// imports 
+
+class Employee {
+  String id;
+  String name;
+  LocalDate joiningDate;
+  String pastEmployment;
+
+  // standard constructor, getters, setters
+}
+```
+
+Mostly the `id` field is going to be a non-nullable value. So, to avoid any potential `NullPointerException` we can mark this field as [`NonNull`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNull.html):
+
+
+```java
+class Employee {
+
+  @NonNull 
+  String id;
+
+  //...
+}
+```
+Now, if we accidentally try to set the value of `id` as null anywhere in the code, the IDE will show a compilation warning:
+
+{{% image alt="IDE warning for NonNull" src="images/posts/spring-boot-null-safety-annotations/nonnull-ide-warning.png" %}}
+
+Pretty awesome, isn't it?
+
+**The `NonNull` annotation can be used at the method, parameter, or field level.** 
```

Comment:
Could you maybe add a [table](https://reflectoring.io/contribute/writing-guide/#tables) to each section that explains what happens when you put the respective annotation on a method, parameter, field, package?

A table is also a nice way to break up the text a bit :)

Something like:

| Annotated element | Effect |
|--------------------------|---------|
| field | Shows a warning when the field is null |
| parameter | ... |

Ideally also show a small code example for each of these cases, just to make it extra clear to the reader how to use it.

Reply:
Great suggestion! I have added a table at the end of each section to summarize the annotation behavior. Is it OK if we skip adding a code snippet in the table to make it concise (because for a couple of annotations it would require an extra clarification)?